File: backend/app/domains/persons/age.py
"""A person's age, and whether it fits an age restriction."""

from datetime import date


def age_on(date_of_birth: date, reference: date) -> int:
    """Whole years completed on ``reference``."""
    age = reference.year - date_of_birth.year
    if (reference.month, reference.day) < (date_of_birth.month, date_of_birth.day):
        age -= 1
    return age


def age_restriction_problem(
    date_of_birth: date | None,
    min_age: int | None,
    max_age: int | None,
    reference: date,
) -> str | None:
    """Why this birth date does not satisfy the restriction, or None if it does.

    An unknown birth date does not pass a stated restriction: waiving the limit
    because the club lacks the data is how an adult got into a youth camp
    (#291). Returns ``"birth_date_required"``, ``"below_min_age"`` or
    ``"above_max_age"``.
    """
    if min_age is None and max_age is None:
        return None
    if date_of_birth is None:
        return "birth_date_required"
    age = age_on(date_of_birth, reference)
    if min_age is not None and age < min_age:
        return "below_min_age"
    if max_age is not None and age > max_age:
        return "above_max_age"
    return None

File: frontend/lib/age.ts
/**
 * Mirrors backend `app/domains/persons/age.py`, so the UI hides or flags what
 * the API would refuse.
 */

/** Whole years completed today, or null when the birth date is unknown. */
export function ageOn(dateOfBirth: string | null | undefined, today = new Date()): number | null {
  const match = dateOfBirth?.match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (!match) return null;
  const [year, month, day] = match.slice(1).map(Number);
  let age = today.getFullYear() - year;
  const beforeBirthday =
    today.getMonth() + 1 < month || (today.getMonth() + 1 === month && today.getDate() < day);
  if (beforeBirthday) age -= 1;
  return age;
}

export type AgeProblem = "birth_date_required" | "below_min_age" | "above_max_age";

/** Why this birth date does not satisfy an age range; null when it does. */
export function ageRestrictionProblem(
  dateOfBirth: string | null | undefined,
  minAge: number | null | undefined,
  maxAge: number | null | undefined,
): AgeProblem | null {
  if (minAge == null && maxAge == null) return null;
  const age = ageOn(dateOfBirth);
  if (age === null) return "birth_date_required";
  if (minAge != null && age < minAge) return "below_min_age";
  if (maxAge != null && age > maxAge) return "above_max_age";
  return null;
}
